consensus/istanbul/backend/announce.go

```diff
@@ -17,405 +17,671 @@
+// The announce thread does 3 things
+// 1) It will poll to see if this node should send an announce once a minute
+// 2) If it should announce, then it will periodically gossip an announce message
+// 3) Regardless of whether it should announce, it will periodically ask it's peers for their announceVersions set, and update it's own announce cache accordingly
+func (sb *Backend) announceThread() {
+	logger := sb.logger.New("func", "announceThread")
+
+	sb.announceThreadWg.Add(1)
+	defer sb.announceThreadWg.Done()
+
+	// Create a ticker to poll if istanbul core is running and check if this is a registered/elected validator.
+	// If both conditions are true, then this node should announce.
+	checkIfShouldAnnounceTicker := time.NewTicker(5 * time.Second)
+
+	// Create a ticker to check peers' announce versions one every 10 minutes
+	announceVersionsCheckTicker := time.NewTicker(10 * time.Minute)
+
+	// Create all the variables needed for the periodic gossip
+	var announceGossipTicker *time.Ticker
+	var announceGossipTickerCh <-chan time.Time
+	var announceGossipFrequencyState AnnounceGossipFrequencyState
+	var currentAnnounceGossipTickerDuration time.Duration
+	var numGossipedMsgsInHighFreqAfterFirstPeerState int
+
+	for {
+		select {
+		case <-checkIfShouldAnnounceTicker.C:
+			logger.Trace("Checking if this node should announce it's enode")
+
+			shouldAnnounce, err := sb.shouldGenerateAndProcessAnnounce()
+			if err != nil {
+				logger.Warn("Error in checking if should announce", err)
+				break
+			}
+
+			if shouldAnnounce && announceGossipTickerCh == nil {
+				// Immediately gossip an announce
+				go sb.gossipAnnounce()
+
+				if sb.config.AnnounceAggressiveGossipOnEnablement {
+					announceGossipFrequencyState = HighFreqBeforeFirstPeerState
 
-func (ar *announceRecord) String() string {
-	return fmt.Sprintf("{DestAddress: %s, EncryptedEnodeURL length: %d}", ar.DestAddress.String(), len(ar.EncryptedEnodeURL))
+					// Send an announce message once a minute
+					currentAnnounceGossipTickerDuration = 1 * time.Minute
+
+					numGossipedMsgsInHighFreqAfterFirstPeerState = 0
+				} else {
+					announceGossipFrequencyState = LowFreqState
+					currentAnnounceGossipTickerDuration = time.Duration(sb.config.AnnounceGossipPeriod) * time.Second
+				}
+
+				// Enable periodic gossiping by setting announceGossipTickerCh to non nil value
+				announceGossipTicker = time.NewTicker(currentAnnounceGossipTickerDuration)
+				announceGossipTickerCh = announceGossipTicker.C
+				logger.Trace("Enabled periodic gossiping of announce message")
+			} else if !shouldAnnounce && announceGossipTickerCh != nil {
+				// Disable periodic gossiping by setting announceGossipTickerCh to nil
+				announceGossipTicker.Stop()
+				announceGossipTickerCh = nil
+				logger.Trace("Disabled periodic gossiping of announce message")
+			}
+
+			// Use this timer to also prune all announce related data structures.
+			if err := sb.pruneAnnounceDataStructures(); err != nil {
```

Comment:
Pruning announce data structures could be expensive due to the need for EVM data (and less so because it loops over all the cached data.). It may not be wise to do that every 5 seconds. I might suggest pruning the table as part of the send process, but maybe that is too infrequent.

Reply:
I placed it in the "send" timer.  This pruning can happen relatively infrequently.